File: NeuroevolutionaryAgents/src/neuralNetwork.py
from configNeuroevolutionaryAgents import *

from genome import *
from neuron import *

class NeuralNetwork:
    def __init__(self):
        self.LayerNumbersToNodeIDs = []
        self.Neurons = {}
        self.InputIDs = []
        self.OutputIDs = []
        self.Actions = {}
        
    def GenerateFromGenome(self, genome : Genome):
        # add node genes to neurons
        for nodeGeneId, nodeGene in genome.GetNodeGenes().items():
            neuron = Neuron()
            neuron.SetNetworkNodeType(NetworkNodeType.HIDDEN)
            if nodeGene.GetNetworkNodeType() == NetworkNodeType.INPUT:
                self.InputIDs.append(nodeGene.GetNodeId())
                neuron.AddInputConnection(0) # id 0 because the input comes from "before" the input layer
                neuron.SetNetworkNodeType(NetworkNodeType.INPUT)
            elif nodeGene.GetNetworkNodeType() == NetworkNodeType.OUTPUT:
                self.OutputIDs.append(nodeGene.GetNodeId())
                neuron.SetNetworkNodeType(NetworkNodeType.OUTPUT)
            neuron.SetLayerNumber(nodeGene.GetLayerNumber())
            self.LayerNumbersToNodeIDs.append((nodeGene.GetLayerNumber(), nodeGene.GetNodeId()))
            self.Neurons[nodeGene.GetNodeId()] = neuron
        # sort neuron layer numbers to IDs mapping so that network layers are calculated sequentially
        self.LayerNumbersToNodeIDs = sorted(self.LayerNumbersToNodeIDs, key=lambda tup: tup[0])
        # add connection genes to neurons
        for connectionGeneId, connectionGene in genome.GetConnectionGenes().items():
            if connectionGene.IsEnabled() == False:
                continue
            # for neuron that leads in to connection, add output connection; out node ID and weight
            self.Neurons[connectionGene.GetInNode()].AddOutputConnection(connectionGene.GetOutNode(), connectionGene.GetWeight())
            # for neuron that recieves from connection, add input connection; in node ID
            self.Neurons[connectionGene.GetOutNode()].AddInputConnection(connectionGene.GetInNode())
            
    def Calculate(self, Senses : dict):
        # feed input neurons with initial input values from the senses
        for senseKey, senseValue in Senses.items():
            self.Neurons[senseKey].FeedInput(0, senseValue) # feed input neurons from Senses, ID 0 because the input comes from "before" the input layer
        # loop through the rest of the network in order of the order number of the neurons
        for layerNumber, nodeId in self.LayerNumbersToNodeIDs:
            # calculate output
            self.Neurons[nodeId].CalculateOutput()
            if len(self.Neurons[nodeId].GetOutputWeights()) > 0: # if neuron has outputs
                # feed the next neurons
                for outputID, outputWeight in self.Neurons[nodeId].GetOutputWeights().items():
                    self.Neurons[outputID].FeedInput(nodeId, outputWeight * self.Neurons[nodeId].GetOutputValue())
        # get the output from output neurons into network output (Actions)
        for outputID in self.OutputIDs:
            self.Actions[outputID] = self.Neurons[outputID].GetOutputValue()
        # return network output (Actions)
        return self.Actions

File: NeuralLife/NeuralLife/neuralNetwork.cpp
#include "neuralNetwork.hpp"

NeuralNetwork::NeuralNetwork()
{
	// init actions map
	for (size_t i = n_Senses + 1; i <= n_Senses + n_Actions; i++) {
		Actions.insert({ i, 0.f });
	}
}

NeuralNetwork::~NeuralNetwork()
{

}

void NeuralNetwork::GenerateFromGenome(Genome genome)
{
	// add node genes to neurons
	for (auto& nodeGene : genome.GetNodeGenes()) {
		Neuron neuron;
		neuron.SetType(HIDDEN);
		if (nodeGene.second.GetType() == NetworkNodeType::INPUT) {
			InputIDs.push_back(nodeGene.second.GetId());
			neuron.AddInputConnection(0); // id 0 because the input comes from "before" the input layer
			neuron.SetType(INPUT);
		}
		else if (nodeGene.second.GetType() == NetworkNodeType::OUTPUT) {
			OutputIDs.push_back(nodeGene.second.GetId());
			neuron.SetType(OUTPUT);
		}
		neuron.SetLayerNumber(nodeGene.second.GetLayerNumber());
		LayerNumbersToIDs.push_back({ nodeGene.second.GetLayerNumber(), nodeGene.second.GetId() });
		Neurons.insert({ nodeGene.second.GetId(), neuron });
	}
	// sort neuron layer numbers to IDs mapping so that network layers are calculated sequentially
	std::sort(LayerNumbersToIDs.begin(), LayerNumbersToIDs.end());
	// add connection genes to neurons
	for (auto& connectionGene : genome.GetConnectionGenes()) {
		if (connectionGene.second.IsEnabled() == false) {
			continue;
		}
		// for neuron that leads in to connection, add output connection; out node ID and weight
		Neurons.at(connectionGene.second.GetInNode()).AddOutputConnection(connectionGene.second.GetOutNode(), connectionGene.second.GetWeight());
		// for neuron that recieves from connection, add input connection; in node ID
		Neurons.at(connectionGene.second.GetOutNode()).AddInputConnection(connectionGene.second.GetInNode());
	}
}

std::map<size_t, float> NeuralNetwork::Calculate(std::map<size_t, float> Senses)
{
	// feed input neurons with initial input values from the senses
	for (auto& sense : Senses) {
		Neurons.at(sense.first).FeedInput(0, sense.second); // feed input neurons from Senses, ID 0 because the input comes from "before" the input layer
	}
	// loop through the rest of the network in order of the order number of the neurons
	for (auto& layerNumberToID : LayerNumbersToIDs) {
		// calculate output
		Neurons.at(layerNumberToID.second).CalculateOutput();
		if (Neurons.at(layerNumberToID.second).GetOutputWeights().size() > 0) { // if neuron has outputs
			// feed the next neurons
			for (auto& output : Neurons.at(layerNumberToID.second).GetOutputWeights()) {
				Neurons.at(output.first).FeedInput(layerNumberToID.second, output.second * Neurons.at(layerNumberToID.second).GetOutputValue());
			}
		}
	}
	// get the output from output neurons into network output (Actions)
	for (size_t i = n_Senses + 1; i <= n_Senses + n_Actions; i++) {
		Actions[i] = Neurons.at(i).GetOutputValue();
	}
	// return network output (Actions)
	return { Actions };
}